presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java

```diff
@@ -348,14 +395,38 @@ public Type getType()
 
         public void processPage(GroupByIdBlock groupIds, Page page)
         {
-            if (step.isInputRaw()) {
+            if (!this.orderByChannels.isEmpty()) {
+                checkArgument(step == Step.SINGLE && pageIndex.isPresent(), "ORDER BY in aggregations is supported only for single step aggregations");
+                pageIndex.ifPresent(value -> value.addPage(page));
+            }
+            else if (step.isInputRaw()) {
                 aggregation.addInput(groupIds, page);
             }
             else {
                 aggregation.addIntermediate(groupIds, page.getBlock(intermediateChannel));
             }
         }
 
+        public void processOrderBy()
+        {
+            if (this.pageIndex.isPresent()) {
+                checkArgument(groupByHash.isPresent(), "No groupByHash");
+                PagesIndex pageIndex = this.pageIndex.get();
+                pageIndex.sort(orderByChannels, orderings);
+                int currentPosition = 0;
+                while (true) {
+                    pageBuilder.reset();
+                    currentPosition = pageIndex.buildPage(currentPosition, orderByOutputChannels, pageBuilder);
```

Comment:
I did the same thing as in OrderByOperator. I don't really get what the problem is. If PagesIndex provides a function buildPage which allows you to select arbitrary channels to construct pages out of, why this is the wrong way to do things? We are constructing new pages not using all input channels. Do we want to put "output channels" as a parameter of PagesIndex? Otherwise how do you move this inside PagesIndex?